src/librustc_llvm/build.rs

```diff
@@ -279,7 +279,11 @@ fn main() {
             let path = PathBuf::from(s);
             println!("cargo:rustc-link-search=native={}",
                      path.parent().unwrap().display());
-            println!("cargo:rustc-link-lib=static={}", stdcppname);
+            if target.contains("windows") {
+                println!("cargo:rustc-link-lib=static-nobundle={}", stdcppname);
```

Comment:
Yes, I think quite nasty or I'm just misunderstanding.

First let's take a look at static libstdc++:
```
nm /mingw64/lib/gcc/x86_64-w64-mingw32/9.2.0/libstdc++.a | grep pthread_mutex_init
                 U __imp_pthread_mutex_init
                 U pthread_mutex_init
                 U pthread_mutex_init
                 [..]
```
There are references to both static and dynamic symbols!

Now using `static` failed undefined with 2 undefined symbols: `__imp_pthread_mutex_{destroy,init}`. I honestly have no idea why picked dynamic symbols, though I haven't chased that rabbit.
It could be mitigated by dynamically linking to winpthread library (by removing `static` in line [298](https://github.com/rust-lang/rust/pull/65911/files/c814d88104a2d8221a59c834b8cb0c408f8d5215#diff-b0767d8366aa7e0a83a847d53032dfd4R298)), windows-gnu ships `libwinpthread-1.dll` anyway because. But I felt it's just replacing one problem with another.

With some guidance I could try to investigate why picks dynamic pthread symbols when using `static` but `static-nobundle` seems to fit here (I asked for try build in OP to be sure).

Reply:
I had the same issue in #65646, with a similar workaround: https://github.com/rust-lang/rust/pull/65646/commits/a505735af618b7e61cbf62b2643d4a701f7ec692